packages/react-reconciler/src/ReactMemoCache.js

```diff
@@ -0,0 +1,77 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import type {MemoCache} from 'shared/ReactTypes';
+import type {Dispatcher} from './ReactInternalTypes';
+
+import ReactSharedInternals from 'shared/ReactSharedInternals';
+const {ReactCurrentDispatcher} = ReactSharedInternals;
+
+export const MEMO_CACHE_UNSET_SENTINEL: symbol = Symbol.for(
+  'react.usememocache_sentinel',
+);
+
+/**
+ * Implementation of `useMemoCache()` return type
+ */
+export default class MemoCacheImpl implements MemoCache {
+  data: Array<any>;
+
+  constructor(data: Array<any>) {
+    this.data = data;
+  }
+
+  get size(): number {
+    return this.data.length;
+  }
+
+  has(index: number): boolean {
+    if (index >= this.data.length) {
+      throw new Error('Memo cache index of out bounds');
+    }
```

Comment:
I think we can skip the bounds checks as something would need to be pretty broken in the compiler to create those and we don't intend this to be used directly.